instance/models/server.py

```diff
@@ -316,18 +319,23 @@ def update_status(self):
 
         # First check if it makes sense to update the current status.
         # This is not the case if we can not interact with the server:
-        if self.status == Status.BuildFailed:
+        if self.status in (Status.BuildFailed, Status.Terminated):
+            return self.status
+        try:
+            os_server = self.os_server
+        except (requests.RequestException, novaclient.exceptions.ClientException):
+            self.logger.debug('Could not reach the OpenStack API')
+            if self.status not in (Status.BuildFailed, Status.Terminated, Status.Pending):
+                self._status_to_unknown()
             return self.status
-
-        os_server = self.os_server
         self.logger.debug('Updating status from nova (currently %s):\n%s', self.status, to_json(os_server))
 
-        if self.status == Status.Building:
+        if self.status in (Status.Building, Status.Unknown):
             self.logger.debug('OpenStack: loaded="%s" status="%s"', os_server._loaded, os_server.status)
             if os_server._loaded and os_server.status == 'ACTIVE':
                 self._status_to_booting()
 
-        elif self.status == Status.Booting and self.public_ip and is_port_open(self.public_ip, 22):
+        if self.status in (Status.Booting, Status.Unknown) and self.public_ip and is_port_open(self.public_ip, 22):
             self._status_to_ready()
```

Comment:
These 2 `if` statements allow for transitions from `Unknown` to `Booting`, and from `Unknown` to `Ready`, but not from `Unknown` to `Building`.
 
Please add a 3rd `if` statement to catch this case:

```
if self.status == Status.Unknown:
    self._status_to_building()
```